File: backend/backend/src/DeviceComponent/Common/DTOs/DeviceChangedMessage.cs
using System.Text.Json.Serialization;
using backend.WebSocketComponent.Common.Api.DTOs;

namespace backend.DeviceComponent.Common.DTOs;

public class DeviceChangedMessage : ITypedMessage
{
    public static string TypeString => "device-changed";

    [JsonIgnore]
    public string InstanceTypeString => TypeString;

    [JsonPropertyName("action")]
    public required string Action { get; set; } // "added" or "removed"

    [JsonPropertyName("deviceInfo")]
    public required DeviceInfo Device { get; set; }
}

public class DeviceInfo
{
    [JsonPropertyName("uuid")]
    public required Guid Uuid { get; set; }

    [JsonPropertyName("displayName")]
    public required string DisplayName { get; set; }
}

File: frontend/src/types/device/DeviceChangedMessage.ts
import { ITypedMessage } from "../ITypedMessage";
import { MessageType } from "../MessageType";
import { DeviceStatus } from "./DeviceStatus";

export interface DeviceInfo {
    uuid: string;
    displayName: string;
    status: DeviceStatus;
}

export class DeviceChangedMessage implements ITypedMessage {
    public readonly type = MessageType.DEVICE_CHANGED;
    public msg: {
        action: "added" | "removed" | "renamed";
        deviceInfo: DeviceInfo;
    };

    public constructor(msg: {
        action: "added" | "removed" | "renamed";
        deviceInfo: DeviceInfo;
    }) {
        this.msg = msg;
    }
}
